mir: separate constants from constant data

zerbina opened the pull request:
## Summary

In the mid- and backend, there's now a clear distinction being made
between *constants* and *constant data*. A *constant* (i.e., a `const`)
links a name/interface with *constant data*, while *constant data* is
just the raw data.

Storage of constant data is now also decoupled from `TSym.ast`,
preparing for the introduction of a dedicated constant data IR.

## Details

#### General architecture

* same as before, constant data is represented by a `PNode` AST
* every complete `PNode` AST representing constant data is identified
  by the new `DataId`
* all constant data the mid- and backend know about is stored in
  `MirEnv` (in a `DataTable`)
* the association between a `DataId` and the data is *bi-directional*
  (a `DataId` maps to a constant expression and a constant expression
  maps to a `DataId`)
* each user-defined constant (`ConstId`) is associated with a `DataId`
* when discovering user-defined constants, `backends.process` and
  `backends.discover` register the data with the environment and link
  it with the constant

#### Rationale

Separating constants from their data allows two constants to share the
same `PNode` AST internally (at least in the backend). It also makes it
easy for the mid-end to create new constant data (e.g., by lifting
constant expression from procedure bodies) itself -- no name nor
interface has to be associated with the data in this case.

#### Implementation

* `DataTable` is a customized version of `BiTable`
* the `PNode` ASTs representing the constant expression are compared by
  their structure
* using structural comparison means that `Obj(a: 0)` and `Obj(b: 0)`
  are treated as being different, even though they represent the same
  data. While not critical, a normalization pass is going be needed
  here, eventually
* the `DataTable` also supports the basic rewind mechanism employed by
  `MirEnv`
* as an interim solution, a separate table is used for mapping
  constants to their body (`MirEnv.bodies`)

#### Code generation

* instead of querying `TSym.ast`, `cgen` and `jsgen` lookup the body
  for constant through `MirEnv`
* constant data in the VM maps directly to the `DataTable` now (that
  is, each item in `DataTable` maps to a complex VM constant). `vmgen`,
  `vmjit`, and `vmbackend` are adjusted accordingly

zerbina committed:
mirtrees: introduce `DataId`

It's a new ID type identifying the *content* of constants.

zerbina committed:
mirenv: implement the data table structure

A custom `BiTable` implementation (`DataTable`) is used for
associating a `DataId` with an expression and vice versa. An instance
of the table is stored in the `MirEnv`.

The data table is also integrated into the rewind mechanism.

zerbina committed:
jsgen: update code-gen for definition of constants

The expression has to be taken from the data table now, not from the
`sym.ast` field.

zerbina committed:
backends: associate user-defined constants with their data

- user-defined constants are associated with a `DataId` (i.e., the ID
  of their content) through the `MirEnv.bodies` map
- `discover` and `process` add the constants' content to the data
  table and register the mapping from the constant to its body

zerbina committed:
cgen: update code-gen for constants

The expression has to be taken from the data table now, not from the
`sym.ast` field.

zerbina committed:
vmgen: serialize constant *data* instead of constants

The difference is somewhat subtle at the moment: constant data is
the actual data, whereas constants associate a name/interface with
constant data.

As a consequence, `DataId` replaces `ConstId` for the link index.

The difference between user- and compiler-defined constants doesn't
matter to the VM -- the VM only cares about the underlying data.

zerbina committed:
vmgen: make sure the constants' type is mapped

Since constant expression AST goes through unification now, the type
assigned to the AST might be a different `PType` than that used for the
`mnkConst` node.

`vmbackend` looks up the type assigned to the AST, so `vmgen` has to
tentatively map and cache that type too.

zerbina commented on compiler/backend/backends.nim:
```diff
@@ -879,8 +880,10 @@ iterator discover*(env: var MirEnv, progress: EnvCheckpoint
   for id, it in since(env.constants, progress.consts):
     # first discover and report the procedures referenced by the constant's
-    # data, allowing
+    # data. This ensures that the callsite can rely on all the constant's
+    # dependencies existing in the environment
```
There's no change in behaviour here, I just forgot to finish the sentence in the original PR.

saem reviewed (approved):
LGTM

Is the idea that eventually `DataTable` might get some out into a small module of its own?

zerbina commented:
> Is the idea that eventually `DataTable` might get some out into a small module of its own?

Yep, that's the plan, although thinking about it again, it might be a better idea to move it into its own module already. I often tend to put too many related-in-usage things into the same module, which is often detrimental in the longer term.

saem commented:
> Yep, that's the plan

Sounds good.

> although thinking about it again, it might be a better idea to move it into its own module already. I often tend to put too many related-in-usage things into the same module, which is often detrimental in the longer term.

Up to you, I think it's fine to merge. I personally think modules are decision encapsulation boundaries, so I don't get too fussed unless the public interface is starting to get ugly/too big. Could just as well pull it out in the next PR.

zerbina committed:
move `DataTable` implementation to dedicated module

It doesn't rely on `MirEnv`, so there's no reason to have both in the
same module.

zerbina commented:
/merge

I've moved the `DataTable` to its own module.
